tools/cbmc/proofs/HTTP/global_state_HTTP.c

```diff
@@ -123,103 +183,193 @@ int is_valid_NetworkInterface(IotNetworkInterface_t *netif) {
   if (pRequestHandle) {
     uint32_t len;
-    pRequestHandle->pHttpsResponse = newIotResponseHandle();
-    pRequestHandle->pHttpsConnection = newIotConnectionHandle();
-    pRequestHandle->pBody = malloc(len);
-    pRequestHandle->pHeaders = ((_reqHandle_t*)pRequestHandle)->data;
-    pRequestHandle->pHeadersCur = pRequestHandle->pHeaders;
-    pRequestHandle->pHeadersEnd = pRequestHandle->pHeaders + sizeof(((_reqHandle_t*)pRequestHandle)->data);
-    pRequestHandle->pConnInfo = newConnectionInfo();
+    pRequestHandle->pHttpsResponse = allocate_IotResponseHandle();
+    pRequestHandle->pHttpsConnection = allocate_IotConnectionHandle();
+    pRequestHandle->pBody = safeMalloc(len);
+    pRequestHandle->pConnInfo = allocate_IotConnectionInfo();
   }
   return pRequestHandle;
 }
+
+int is_valid_IotRequestHandle(IotHttpsRequestHandle_t pRequestHandle) {
+  int required =
+    __CPROVER_same_object(pRequestHandle->pHeaders,
+			  pRequestHandle->pHeadersCur) &&
+    // ???: Does this overconstrain End?
+    __CPROVER_same_object(pRequestHandle->pHeaders,
+			  pRequestHandle->pHeadersEnd);
+  if (!required) return 0;
+
+  int valid_headers =
+    pRequestHandle->pHeaders == ((_reqHandle_t*)pRequestHandle)->data;
+  int valid_order =
+    pRequestHandle->pHeaders <= pRequestHandle->pHeadersCur &&
+    pRequestHandle->pHeadersCur <=  pRequestHandle->pHeadersEnd;
+  int valid_body =
+    pRequestHandle->pBody != NULL;
+  return
+    valid_headers &&
+    valid_order &&
+    valid_body &&
+    // valid_order and short circuit evaluation prevents integer overflow
+    pRequestHandle->pHeadersEnd - pRequestHandle->pHeaders <= USER_DATA_SIZE;
+}
+
+/****************************************************************
+ * IotHttpsRequestInfo constructor
+ * This is currently unusued and untested.
+ ****************************************************************/
+
 /* Creates a Request Info and assigns memory accordingly. */
-IotHttpsRequestInfo_t * newIotRequestInfo() {
-  IotHttpsRequestInfo_t * pReqInfo = safeMalloc(sizeof(IotHttpsRequestInfo_t));
+IotHttpsRequestInfo_t * allocate_IotRequestInfo() {
+  IotHttpsRequestInfo_t * pReqInfo
+    = safeMalloc(sizeof(IotHttpsRequestInfo_t));
   if(pReqInfo) {
-    uint32_t bufferSize;
-    uint32_t hostNameLen;
-    __CPROVER_assume(bufferSize >=0 && bufferSize <= requestUserBufferMinimumSize);
-    pReqInfo->userBuffer.bufferLen = bufferSize;
-    pReqInfo->userBuffer.pBuffer = safeMalloc(bufferSize);
-    __CPROVER_assume(hostNameLen >= 0 && hostNameLen <= IOT_HTTPS_MAX_HOST_NAME_LENGTH + 1);
-    pReqInfo->pHost = safeMalloc(hostNameLen);
+    pReqInfo->userBuffer.pBuffer = safeMalloc(pReqInfo->userBuffer.bufferLen);
+    pReqInfo->pHost = safeMalloc(pReqInfo->hostLen);
   }
   return pReqInfo;
 }
 
+int is_valid_IotRequestInfo(IotHttpsRequestInfo_t * pReqInfo) {
+  return
+    // ???: Should the global minimum size really be an upper bound?
```

Comment:
No it should not actually. I did not catch that before.

Reply:
This is unusued, let's just delete the allocator and is_valid for this type.

Reply:
That sounds good to me.

Reply:
That sounds good to me.